# **Aplikasi Aljabar Linier**

### Definisi
PageRank adalah sebuah algoritma, yang diperkenalkan oleh Larry Page, yang digunakan untuk menentukan peringkat halaman web dan sekarang digunakan di berbagai bidang. Algoritma ini memulai dari masalah kompleks dan berakhir dengan solusi sederhana, menggunakan dasar aljabar dan Rantai Markov. PageRank menggambarkan web sebagai grafik berarah, dengan node sebagai halaman web dan tepi sebagai tautan. Pentingnya halaman ditentukan oleh jumlah tautan yang masuk dan distribusi kepentingan yang merata ke semua halaman yang ditautkan. Solusi untuk menentukan peringkat halaman melibatkan Rantai Markov dan distribusi stasioner, yang menyatu setelah langkah tak terhingga, menghasilkan peringkat halaman yang stabil.

## Cara Kerja Dari Algoritma PageRank
1. **Membuat Grafik Berarah**:
   - Pertama, kita menggambarkan web sebagai grafik berarah, dengan setiap halaman web sebagai simpul (node) dan tautan antar halaman sebagai tepi (edge).
   - Setiap halaman memiliki skor awal yang sama.

2. **Menghitung Skor Awal**:
   - Setiap halaman diberi skor awal, biasanya 1/N (dengan N adalah jumlah total halaman).

3. **Iterasi PageRank**:
   - Lakukan iterasi untuk menghitung skor PageRank hingga konvergensi (stabil).
   - Setiap iterasi melibatkan menghitung skor baru untuk setiap halaman berdasarkan tautan masuk dan skor halaman lain.

4. **Faktor Damping**:
   - Faktor damping (biasanya sekitar 0,85) memastikan bahwa tidak semua skor tautan masuk langsung ditambahkan.
   - Rumus PageRank setelah faktor damping:
     $$ PR(A) = (1-d) + d \left( \frac{PR(T1)}{C(T1)} + \ldots + \frac{PR(Tn)}{C(Tn)} \right) $$

5. **Iterasi Lanjutan**:
   - Lakukan iterasi hingga skor PageRank konvergen (tidak berubah signifikan).

6. **Hasil Akhir**:
   - Halaman dengan skor PageRank tertinggi dianggap lebih penting.



  $$r_j = \sum\limits_{i=1}\frac{r_i}{d_i}$$

  $
  r_0 = \frac{r_4}{3}\\
  r_1 = \frac{r_2}{2}+\frac{r_4}{3}+r_3\\
  r_2 = \frac{r_0}{3}+\frac{r_4}{3}\\
  r_3 = \frac{r_2}{2}+\frac{r_0}{3}\\
  r_4 = \frac{r_0}{3}+r_1\\
  $

  $A = \begin{bmatrix}
  0 & 0 & 0 & 0 & \frac{1}{3}\\
  0 & 0 & \frac{1}{2} & 1 & \frac{1}{3}\\
  \frac{1}{3} & 0 & 0 & 0 & \frac{1}{3}\\
  \frac{1}{3} & 0 & \frac{1}{2} & 0 & 0\\
  \frac{1}{3} & 1 & 0 & 0 & 0\\
  \end{bmatrix}$
  


  $
  V0 =
  \begin{bmatrix}
  0,2 \\
  0,2 \\
  0,2 \\
  0,2 \\
  0,2 \\
  \end{bmatrix}
  $

  toleransi = 0,01





## **Contoh Secara Manual**

In [1]:
import numpy as np
A = np.array([[0, 0, 0, 0, 1/3],[0, 0, 1/2, 1, 1/3],[1/3, 0, 0, 0, 1/3], [1/3, 0, 1/2, 0, 0], [1/3, 1, 0, 0, 0]])
v0 = np.array([[0.2], [0.2], [0.2], [0.2], [0.2]])

v1 = A @ v0
print(v1)


[[0.06666667]
 [0.36666667]
 [0.13333333]
 [0.16666667]
 [0.26666667]]


In [2]:
v2 = A @ v1
print(v2)

[[0.08888889]
 [0.32222222]
 [0.11111111]
 [0.08888889]
 [0.38888889]]


In [3]:
v3 = A @ v2
print (v3)

[[0.12962963]
 [0.27407407]
 [0.15925926]
 [0.08518519]
 [0.35185185]]


In [4]:
v4 = A @ v3
print (v4)

[[0.11728395]
 [0.28209877]
 [0.16049383]
 [0.12283951]
 [0.31728395]]


In [5]:
v5 = A @ v4
print (v5)

[[0.10576132]
 [0.30884774]
 [0.14485597]
 [0.11934156]
 [0.32119342]]


$$
\begin{align*}
&= (0,10 - 0,11)^2 + (0,30 - 0,28)^2 + (0,14 - 0,16)^2 \\
&\quad + (0,11 - 0,12)^2 + (0,32 - 0,31)^2 \\
&= 0,01^2 + 0,02^2 + (-0,02)^2 + (-0,01)^2 + 0,01^2 \\
&= 0,0001 + 0,0004 + 0,0004 + 0,0001 + 0,0001 \\
&= 0,0011 < 0,01
\end{align*}
$$


In [6]:
a = (0.10 - 0.11)**2
b = (0.30 - 0.28)**2
c = (0.14 - 0.16)**2
d = (0.11 - 0.12)**2
e = (0.32 - 0.31)**2

f = a + b + c + d + e

print(f)

0.0010999999999999981


# **Tugas**

## Contoh Memakai Python

In [22]:
import numpy as np

def perkalianMatrix(A, V, conv=0.01, itr=0):
    # Mengalikan matriks A dengan vektor V
    A_ = A.dot(V)
    
    # Memeriksa apakah konvergen (perubahan kurang dari ambang batas)
    if np.linalg.norm(A_ - V) < conv:
        return A_, itr
    else:
        # Jika belum konvergen, lanjutkan iterasi dengan matriks baru
        return perkalianMatrix(A, A_, conv, itr+1)

# matriks A 
A = np.array([[0, 0, 0, 0, 1/3],
              [0, 0, 1/2, 1, 1/3],
              [1/3, 0, 0, 0, 1/3],
              [1/3, 0, 1/2, 0, 0],
              [1/3, 1, 0, 0, 0]])

# vektor PageRank awal (semua halaman memiliki nilai awal 0.2)
V = np.array([0.2, 0.2, 0.2, 0.2, 0.2])

# Hitung PageRank
hasil, iterasi = perkalianMatrix(A, V)

# Tampilkan hasil
print("Hasil konvergen:")
print(hasil)
print()
print("Jumlah iterasi:", iterasi)
print()

# Urutkan nilai PageRank
hasil_terurut = sorted(hasil, reverse=True)
print("Nilai PageRank yang diurutkan:")
for i, pr in enumerate(hasil_terurut):
    print(f"Halaman {i}: {pr:}")


Hasil konvergen:
[0.11150739 0.293549   0.14974089 0.11342783 0.33177488]

Jumlah iterasi: 7

Nilai PageRank yang diurutkan:
Halaman 0: 0.3317748818777625
Halaman 1: 0.29354900167657366
Halaman 2: 0.149740893156531
Halaman 3: 0.11342783112330437
Halaman 4: 0.11150739216582836


# **Penjelasan**

1. **Matriks Transisi \($A$\)**:
   

   \[ $A = \begin{bmatrix}
   0 & 0 & 0 & 0 & \frac{1}{3} \\
   0 & 0 & \frac{1}{2} & 1 & \frac{1}{3} \\
   \frac{1}{3} & 0 & 0 & 0 & \frac{1}{3} \\
   \frac{1}{3} & 0 & \frac{1}{2} & 0 & 0 \\
   \frac{1}{3} & 1 & 0 & 0 & 0 \\
   \end{bmatrix}$ \]

   Pertama, kita sudah memiliki matriks transisi \($A$\). Dan matriks ini menggambarkan probabilitas perpindahan dari satu halaman ke halaman lain. Dimana matriks ini dibangun dari struktur tautan antara halaman web.

   
2. **Vektor PageRank Awal \(V\)**:
   

   \[ $V = \begin{bmatrix}
   0.2 \\
   0.2 \\
   0.2 \\
   0.2 \\
   0.2 \\
   \end{bmatrix}$ \]


   saya memiliki vektor PageRank awal \($V$\). Vektor ini menggambarkan nilai awal PageRank untuk setiap halaman. Pada awalnya, semua halaman mungkin memiliki nilai yang sama (seperti ini, \($0.2$\)).

   
3. **Perhitungan PageRank**:

   Iterasi dilakukan dengan mengalikan matriks \($A$\) dengan vektor \($V$\). Hasilnya adalah vektor baru \($V'$\):

   \[$ V' = A \cdot V $\]

   \[ $A = \begin{bmatrix}                                             
   0 & 0 & 0 & 0 & \frac{1}{3} \\
   0 & 0 & \frac{1}{2} & 1 & \frac{1}{3} \\                
   \frac{1}{3} & 0 & 0 & 0 & \frac{1}{3} \\
   \frac{1}{3} & 0 & \frac{1}{2} & 0 & 0 \\
   \frac{1}{3} & 1 & 0 & 0 & 0 \\
   \end{bmatrix}$ \] 
   
    $\cdot$

   \[ $V = \begin{bmatrix}
   0.2 \\
   0.2 \\
   0.2 \\
   0.2 \\
   0.2 \\
   \end{bmatrix}$ \]
   
   Setelah itu, periksa apakah perubahan antara \($V'$\) dan \($V$\) kurang dari ambang batas (\($0.01$\)). Jika ya, kita menganggap konvergen dan menghentikan iterasi. Jika tidak, kita lanjutkan iterasi dengan menggunakan \($V'$\) sebagai vektor PageRank baru.

4. **Hasil Konvergen**:

   Hasil konvergen adalah:

   \[ $\text{PageRank} = \begin{bmatrix}
   0.331 \\
   0.293 \\
   0.149 \\
   0.113 \\
   0.331 \\
   \end{bmatrix}$ \]

5. **Nilai PageRank yang Diurutkan**:

   Nilai PageRank yang diurutkan dari yang tertinggi ke terendah:

   - Halaman 1: 0.331
   - Halaman 2: 0.293
   - Halaman 3: 0.149
   - Halaman 4: 0.113
   - Halaman 5: 0.111
